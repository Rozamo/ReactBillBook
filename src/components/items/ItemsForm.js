import React, { useState } from 'react';
import { useHistory } from 'react-router-dom';
import PostData from '../helper/api/PostData';
import loadingGif from '../../images/loading.gif';
import BlueButton from '../bluebutton/BlueButton';

export default function ItemsForm() {
  const history = useHistory();
  const [name, setName] = useState('');
  const [description, setDescription] = useState('');
  const [amount, setAmount] = useState('');
  const currency = 'INR';
  const [isPosting, setIsPosting] = useState(false);

  function isValid() {
    if (!/^[0-9]*.?[0-9]+$/.test(amount)) {
      alert('Amount must be valid');
      return false;
    }
    return true;
  }

  async function handleFormSubmit(event) {
    event.preventDefault();
    if (!isValid()) return;
    setIsPosting(true);
    const newAmount = Number(amount) * 100;
    const dataToPost = {
      name, description, amount: newAmount, currency,
    };
    const responseData = await PostData(dataToPost, 'items');
    if (responseData.statusCode === 400) alert(responseData.error.description);
    else if (responseData.entity === 'item' || responseData.id) {
      setIsPosting(false);
      history.push('/items/list', { submitSuccess: true });
    } else alert(responseData);
    setIsPosting(false);
  }

  function handleContent() {
    if (isPosting) return <img src={loadingGif} alt="Loading...." id="load-img" />;
    return (
      <form
        className="customer-form"
        style={{ width: '50%' }}
        onSubmit={handleFormSubmit}
      >
        <label htmlFor="name">Name</label>
        <input
          value={name}
          type="text"
          name="name"
          onChange={(event) => {
            setName(event.target.value);
          }}
          required
        />
        <label htmlFor="amount">Price</label>
        <input
          value={amount}
          type="number"
          name="amount"
          onChange={(event) => {
            setAmount(event.target.value);
          }}
          required
        />
        <label htmlFor="name">Description</label>
        <textarea
          value={description}
          type="text"
          name="description"
          onChange={(event) => {
            setDescription(event.target.value);
          }}
        />
        <br />
        <BlueButton
          sidebarChoice="items"
          contentChoice="create"
          type="submit"
        />
      </form>
    );
  }

  return (
    <div className="content">
      <div className="top-panel">
        <h1 id="title">New Item</h1>
      </div>
      {handleContent()}
    </div>
  );
}
