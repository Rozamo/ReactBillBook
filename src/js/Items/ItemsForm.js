import React, { useState } from 'react';
import PostData from '../Helper/API/PostData';
import loadingGif from '../../images/loading.gif';
import floppy from '../BlueButton/floppy.png';
import {useHistory} from 'react-router-dom';
import BlueButton from '../BlueButton/BlueButton';

export default function ItemsForm() {
    const history = useHistory();
    const [name, setName] = useState('');
    const [description, setDescription] = useState('');
    const [amount, setAmount] = useState('');
    const currency = 'INR';
    const [isPosting, setIsPosting] = useState(false);
    function isValid() {
        if (!/^[0-9]*.?[0-9]+$/.test(amount)) {
            alert('Amount must be valid');
            return false;
        }
        return true;
    }
    async function handleFormSubmit(event) {
        event.preventDefault();
        if (!isValid())
            return;
        setIsPosting(true);
        const new_amount = Number(amount) * 100;
        const new_obj = {name, description, amount: new_amount, currency};
        const data = await PostData(new_obj, 'items');
        if (data.statusCode === 400)
            alert(data.error.description);
        else if (data.entity === 'customer' || data.id)
            history.push('/items/list');
        else
            alert(data);
        setIsPosting(false);
    }
    function handleContent() {
        if (isPosting)
            return <img src={loadingGif} alt="Loading...." id="load-img"></img>;
        else return <form className="customer-form" style={{width: "50%"}} onSubmit={handleFormSubmit}>
            <label htmlFor="name">Name</label>
            <input value={name} type="text" name="name" onChange={(event) => {setName(event.target.value)}}/>
            <label htmlFor="amount">Price</label>
            <input value={amount} type="number" name="amount" onChange={(event) => {setAmount(event.target.value)}}/>
            <label htmlFor="name">Description</label>
            <textarea value={description} type="text" name="description" onChange={(event) => {setDescription(event.target.value)}}/>
            <br></br>
            <BlueButton sidebarChoice='items' contentChoice='create' type='submit'/>
        </form>;
    }
    return <div className="content">
        <div className="top-panel">
            <h1 id="title">New Item</h1>
        </div>     
        {handleContent()}        
    </div>;
}
