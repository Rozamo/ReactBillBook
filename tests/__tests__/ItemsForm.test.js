import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import ItemsForm from "../../src/components/items/ItemsForm";

let container = null;
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe("Customer form", () => {
  it("should render form", () => {
    act(() => {
      render(<ItemsForm />, container);
    });
    expect(container.innerHTML).toMatchInlineSnapshot(
      `"<div class=\\"content\\"><div class=\\"top-panel\\"><h1 id=\\"title\\">New Item</h1></div><form class=\\"customer-form\\" style=\\"width: 50%;\\"><label for=\\"name\\">Name</label><input type=\\"text\\" name=\\"name\\" value=\\"\\"><label for=\\"amount\\">Price</label><input type=\\"number\\" name=\\"amount\\" value=\\"\\"><label for=\\"name\\">Description</label><textarea type=\\"text\\" name=\\"description\\"></textarea><br><button type=\\"submit\\" id=\\"button\\" value=\\"\\"><img src=\\"test-file-stub\\" id=\\"floppy\\" alt=\\"Save\\">Save Item</button></form></div>"`
    );

    expect(container.textContent).toContain("Name");
    const name = document.querySelectorAll("input")[0];
    const desc = document.getElementsByName("description")[0];

    name.value = "Mouse";
    desc.value = "desc";

    expect(name.value.length).toBe(5);
    expect(desc.value).toBe("desc");
  });
});
