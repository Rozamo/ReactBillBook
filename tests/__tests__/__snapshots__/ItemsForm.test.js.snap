// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Items form should render ItemsForm 1`] = `
<div>
  <div
    class="content"
  >
    <div
      class="top-panel"
    >
      <h1
        id="title"
      >
        New Item
      </h1>
    </div>
    <form
      class="customer-form"
      style="width: 50%;"
    >
      <label
        for="name"
      >
        Name
      </label>
      <input
        name="name"
        type="text"
        value=""
      />
      <label
        for="amount"
      >
        Price
      </label>
      <input
        name="amount"
        type="number"
        value=""
      />
      <label
        for="name"
      >
        Description
      </label>
      <textarea
        name="description"
        type="text"
      />
      <br />
      <button
        id="button"
        type="submit"
        value=""
      >
        <img
          alt="Save"
          id="floppy"
          src="test-file-stub"
        />
        Save Item
      </button>
    </form>
  </div>
</div>
`;

exports[`Items form should render ItemsForm 2`] = `
<div>
  <div
    class="content"
  >
    <div
      class="top-panel"
    >
      <h1
        id="title"
      >
        New Item
      </h1>
    </div>
    <form
      class="customer-form"
      style="width: 50%;"
    >
      <label
        for="name"
      >
        Name
      </label>
      <input
        name="name"
        type="text"
        value="React"
      />
      <label
        for="amount"
      >
        Price
      </label>
      <input
        name="amount"
        type="number"
        value="20000"
      />
      <label
        for="name"
      >
        Description
      </label>
      <textarea
        name="description"
        type="text"
      >
        desc
      </textarea>
      <br />
      <button
        id="button"
        type="submit"
        value=""
      >
        <img
          alt="Save"
          id="floppy"
          src="test-file-stub"
        />
        Save Item
      </button>
    </form>
  </div>
</div>
`;

exports[`Items form should render ItemsForm 3`] = `
<div>
  <div
    class="content"
  >
    <div
      class="top-panel"
    >
      <h1
        id="title"
      >
        New Item
      </h1>
    </div>
    <img
      alt="Loading...."
      id="load-img"
      src="test-file-stub"
    />
  </div>
</div>
`;

exports[`Items form should render ItemsForm 4`] = `
<div>
  <div
    class="content"
  >
    <div
      class="top-panel"
    >
      <h1
        id="title"
      >
        New Item
      </h1>
    </div>
    <form
      class="customer-form"
      style="width: 50%;"
    >
      <label
        for="name"
      >
        Name
      </label>
      <input
        name="name"
        type="text"
        value="React"
      />
      <label
        for="amount"
      >
        Price
      </label>
      <input
        name="amount"
        type="number"
        value="20000"
      />
      <label
        for="name"
      >
        Description
      </label>
      <textarea
        name="description"
        type="text"
      >
        desc
      </textarea>
      <br />
      <button
        id="button"
        type="submit"
        value=""
      >
        <img
          alt="Save"
          id="floppy"
          src="test-file-stub"
        />
        Save Item
      </button>
    </form>
  </div>
</div>
`;
